doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```diff
@@ -0,0 +1,205 @@
+
+### Making things less confusing with distinct function names
+
+So far, `EVP_PKEY_sign()` and friends are only expected to act on the
+pre-computed digest of a message (under the condition that proper flags
+and signature md are specified using functions like
+`EVP_PKEY_CTX_set_rsa_padding()` and `EVP_PKEY_CTX_set_signature_md()`),
+or to act as "primitive" [^1] functions (under the condition that proper
+flags are specified, like `RSA_NO_PADDING` for RSA signatures).
+
+This design proposes an extension to also allow full (not pre-hashed)
+messages to be passed, in a streaming style through an *update* and a
+*final* function.
+
+Discussions have revealed that it is potentially confusing to confound the
+current functionality with streaming style functionality into the same name,
+so this design separates those out with specific init / update / final
+functions for that purpose.  For oneshot functionality, `EVP_PKEY_sign()`
+and `EVP_PKEY_verify()` remain supported, through an alias for the
+application.
+
+[^1]: the term "primitive" is borrowed from [PKCS#1](https://www.rfc-editor.org/rfc/rfc8017#section-5)
+
+### Making it possible to verify with an early signature
+
+There are newer verifying algorithms that need to receive the signature
+before processing the data.  This is particularly important for streaming
+functionality.  This design proposes a mechanism to accomodate this for the
+streaming style functionality.
+
+New public API - API Reference
+------------------------------
+
+### For limited input size / oneshot signing with `EVP_SIGNATURE`
+
+``` C
+int EVP_PKEY_sign_init_ex2(EVP_PKEY_CTX *pctx,
+                           EVP_SIGNATURE *algo,
+                           const OSSL_PARAM params[]);
+```
+
+### For signing a stream with `EVP_SIGNATURE`
+
+``` C
+int EVP_PKEY_sign_message_init(EVP_PKEY_CTX *pctx,
+                               EVP_SIGNATURE *algo,
+                               const OSSL_PARAM params[]);
+int EVP_PKEY_sign_message_update(EVP_PKEY_CTX *ctx,
+                                 const unsigned char *in,
+                                 size_t inlen);
+int EVP_PKEY_sign_message_final(EVP_PKEY_CTX *ctx,
+                                unsigned char *sig,
+                                size_t *siglen);
+#define EVP_PKEY_sign_message(ctx,sig,siglen,tbs,tbslen) \
+    EVP_PKEY_sign(ctx,sig,siglen,tbs,tbslen)
+```
+
+### For limited input size / oneshot verification with `EVP_SIGNATURE`
+
+``` C
+int EVP_PKEY_verify_init_ex2(EVP_PKEY_CTX *pctx,
+                             EVP_SIGNATURE *algo,
+                             const OSSL_PARAM params[]);
+```
+
+### For verifying a stream with `EVP_SIGNATURE`
+
+``` C
+/* Initializers */
+int EVP_PKEY_verify_message_init(EVP_PKEY_CTX *pctx,
+                                 EVP_SIGNATURE *algo,
+                                 const OSSL_PARAM params[]);
+/* Signature setter */
+int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
```

Comment:
Is this name ok?
It doesnt match all the others here (the CTX bit) - should we keep it consistent with the others..

Why doesnt the name have _verify_ in its name if it is only used for verification?

Having a seperate function is good for the case where the signature is needed early. For the other case it causes an unnecessary mem copy since the signature cant be applied via the message_final().
Is this ok?

Reply:
What about calling the function EVP_PKEY_CTX_set0_signature() or EVP_PKEY_verify_message_set0_signature()?

And use octet ref instead of octet string param. Of course that means usage in applications for the case where the message needs to be set before calling the update function(s) will be more complicated. Maybe the extra allocation/memcpy()/free call is not really relevant given the complexity of the signature verification.

Reply:
> Is the EVP_PKEY_CTX_set_signature name ok?
> It doesnt match all the others here (the CTX bit) - should we keep it consistent with the others, even if it is wrong.

I modelled that after already existing patterns, see EVP_PKEY_CTX_set_signature_md() and EVP_PKEY_CTX_set_rsa_padding().

> Why doesnt the name have _verify_ in its name if it is only used for verification?

Because the *thing* that's being set is the signature to verify against.  We don't have a *thing* called "verification", do we?

> Having a separate function is good for the exceptional case where the signature is needed early. For the normal case it causes an unnecessary memcopy since the signature cant be applied via the message_final(). Is this ok?

This was a choice to avoid causing confusion by allowing two ways to set that signature.  What happens if someone gives the signature to verify against using both EVP_PKEY_CTX_set_signature() and EVP_PKEY_verify_message_final() (if we imagine the latter with the added arguments to set the signature that way)?

Reply:
> What about calling the function EVP_PKEY_CTX_set0_signature() or EVP_PKEY_verify_message_set0_signature()?

I could go for either.

> And use octet ref instead of octet string param.

Which means making it another provider function...  well, considering that a signature is mandatory for verification, that makes sense.

Reply:
I prefer EVP_PKEY_verify_message_set0_signature() since it matches the other (wrong) ones :)

Reply:
set0?  set1?